go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go

```diff
@@ -298,12 +335,33 @@ func (c *Controller) expandANPRulePeersToIPSet(peers []*adminNetworkPolicyPeer,
 						// move on to next item in the loop
 						continue
 					}
-					return err
+					return err // we won't hit this TBH because the only error that GetPodIPsOfNetwork returns is podIPsNotFound
 				}
 				peerIPs.Insert(util.StringSlice(podIPs)...)
 				podCache.Insert(pod.Name)
+				// Process NamedPorts if any
+				if len(namedPorts) == 0 {
+					continue
+				}
+				for _, container := range pod.Spec.Containers {
+					for _, port := range container.Ports { // this loop is/might get expensive
+						if port.Name == "" {
+							continue
+						}
+						namedPortReps, ok := namedPorts[port.Name]
+						if !ok {
+							continue
+						}
+						*namedPortReps = append(*namedPortReps, convertPodIPContainerPortToNNPP(port, podIPs)...)
```

Comment:
trying to think why I had to do this, so I think it was because the goal here is to edit the rule.namedPorts inplace, so I needed to use pointers and not pass copies around. So this in place change is what helps me straight sort in one go; https://github.com/ovn-org/ovn-kubernetes/pull/4247/commits/701bf8aa21cd3f95556b68f898fe51a74921e47a#diff-06bcdf751c69ac9530a7498b278b196f0e38cb08ca14546f0208a1bb5a19022aR399

Reply:
I am still not sure I understand, so if didn't use list references, could that loop just be 
```
namedPortRepresentations2 := map[string][][]libovsdbutil.NamedNetworkPolicyPort{}
for i := range namedPortRepresentations2[port.Name] {
	namedPortRepresentations2[port.Name][i] = append(namedPortRepresentations2[port.Name][i], convertPodIPContainerPortToNNPP(port, podIPs)...)
}
```
? I am probably missing something here

Reply:
@npinaeva : sorry I missed this comment in my last push; let me try to explain this better, acknowledging the comment and will look into this later today (just loading back into context after hols)

Reply:
So let me explain what I am doing and then maybe we can see if I can simply this:

step1: when we create an ANP in types.go we clone that ANP into a cache representation; so at this point we have the name of the namedPort stored in our ovnk anpCache: `anpRule.namedPorts[*port.NamedPort] = &[]libovsdbutil.NamedNetworkPolicyPort{}` (let's say I don't create a slice of pointers; then when I reference this array by passing it as a param I'd be passing the a copy of the value which is not desired)

step2: we then in our sync ensureANP function need to expand this namedport into per pod representation. which is this function `convertANPSubjectToLSPs` for ingress and `expandRulePeers` for egress; so in this step what I want to do is to modify rule.namedPorts cache directly by using pointer reference; so what I want to store in `namedPortRepresentations` for mutating is not a copy of the cache/map but reference of the original cache that can be modified in place.

If I don't create the pointer slice of port representations; then when I pass around the cache in the functions I'll be passing by value not by reference so any modifications will be on the copy here and not on the actual cache.

lmk if I made sense and we can huddle and I can try to explain better because I might also be missing something here..

Reply:
nice explanation, thank you! I see that my confusion mostly came from `expandRulePeers`, where this structure is not as required as in `convertANPSubjectToLSPs`. I have another suggestion, which is very similar to your code, but instead of using references to the resolved named ports, it stores indices of ingress rules, which use these ports, that allows to update lists in the original structure:
```
	namedPortRepresentations2 := map[string][]int{}
        for i, ingress := range anp.ingressRules {
		for name, port := range ingress.namedPorts {
			namedPortRepresentations2[name] = append(namedPortRepresentations2[name], i)
		}
	}
...
                                 idxs, ok := namedPortRepresentations2[port.Name]
					if !ok {
						continue
					}
					for _, idx := range idxs {
// here dereferenceing may be removed if we change the type or ingressRule.namedPorts
						*anp.ingressRules[idx].namedPorts[port.Name] = append(*anp.ingressRules[idx].namedPorts[port.Name],
							convertPodIPContainerPortToNNPP(port, podIPs)...)
					}
...
        for _, ingress := range anp.ingressRules {
		for _, port := range ingress.namedPorts {
// sortNamedPorts won't need to dereference a list
			sortNamedPorts(port)
		}
	}
```
I think it may be more explicit about what is being updated, lmk if that makes sense

Reply:
ah I see! yes I think your suggestion will also work, let me try this out in `expandRulePeers` like you mentioned since that is mainly where the confusion is since its not super explicit, so your code might make it more transparent to the readers. thanks Nadia!

Reply:
> nice explanation, thank you! I see that my confusion mostly came from `expandRulePeers`, where this structure is not as required as in `convertANPSubjectToLSPs`.

yea I understand; the subtleness in `expandRulePeers` comes because of `convertANPSubjectToLSPs`

Reply:
> 
> ```
> ...
>         for _, ingress := range anp.ingressRules {
> 		for _, port := range ingress.namedPorts {
> // sortNamedPorts won't need to dereference a list
> 			sortNamedPorts(port)
> 		}
> 	}
> ```
> 
> I think it may be more explicit about what is being updated, lmk if that makes sense

I think till that part it was fine, but I still need to dereference here because port is also pointer to slice... here let me push the new commit I did on top so that you see what I mean..

Reply:
@npinaeva see https://github.com/ovn-org/ovn-kubernetes/pull/4247/commits/db58bd8583262b443f278d93ded345993c45819d commit is that what you had in mind? hoping that looks better?

Reply:
yeah exactly, but I also thought that will allow to change namedPort to a simple list, like so:
```
type gressRule struct {
    namedPorts map[string][]libovsdbutil.NamedNetworkPolicyPort
}
```

Reply:
ah I see, ack let me see if I can remove reference everywhere.. will update that in the same last commit so that its easy for review

Reply:
all right, this one is done, resolving comment. PTAL the last commit.